bundles/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/SerialIoThread.java

```diff
@@ -0,0 +1,297 @@
+import org.eclipse.smarthome.io.transport.serial.SerialPortEvent;
+import org.eclipse.smarthome.io.transport.serial.SerialPortEventListener;
+import org.openhab.binding.upb.handler.UPBIoHandler.CmdStatus;
+import org.openhab.binding.upb.internal.message.MessageBuilder;
+import org.openhab.binding.upb.internal.message.UPBMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Event loop for serial communications. Handles sending and receiving UPB messages.
+ *
+ * @author Marcus Better - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class SerialIoThread extends Thread implements SerialPortEventListener {
+    private static final int WRITE_QUEUE_LENGTH = 128;
+    private static final int ACK_TIMEOUT_MS = 500;
+    private static final byte[] ENABLE_MESSAGE_MODE_CMD = { 0x17, 0x70, 0x02, (byte) 0x8e, 0x0d };
+
+    private final Logger logger = LoggerFactory.getLogger(SerialIoThread.class);
+    private final byte[] buffer = new byte[512];
+    private final MessageListener listener;
+    // Single-threaded executor for writes that serves to serialize writes.
+    private final ExecutorService writeExecutor = new ThreadPoolExecutor(1, 1, 30, TimeUnit.SECONDS,
+            new LinkedBlockingQueue<>(WRITE_QUEUE_LENGTH), new NamedThreadFactory("upb-serial-writer", true));
+    private final SerialPort serialPort;
+
+    private int bufferLength = 0;
+    private volatile @Nullable WriteRunnable currentWrite;
+    private volatile boolean done;
+
+    public SerialIoThread(final SerialPort serialPort, final MessageListener listener) {
+        this.serialPort = serialPort;
+        this.listener = listener;
+        setName("upb-serial-reader");
+        setDaemon(true);
+    }
+
+    @Override
+    public void serialEvent(final SerialPortEvent event) {
+        try {
+            logger.trace("RXTX library CPU load workaround, sleep forever");
+            Thread.sleep(Long.MAX_VALUE);
+        } catch (final InterruptedException e) {
+            // ignore
+        }
+    }
+
+    @Override
+    public void run() {
+        // RXTX serial port library causes high CPU load
+        // Start event listener, which will just sleep and slow down event loop
+        try {
+            serialPort.addEventListener(this);
+        } catch (final TooManyListenersException e) {
+            logger.warn("serial port setup failed", e);
+            return;
+        }
+        serialPort.notifyOnDataAvailable(true);
+
+        final byte[] buffer = new byte[256];
+        try (final InputStream in = serialPort.getInputStream()) {
+            if (in == null) {
+                // should never happen
+                throw new IllegalStateException("serial port is not readable");
+            }
+            enterMessageMode();
+            int len;
+            while (!done && (len = in.read(buffer)) >= 0) {
+                addData(buffer, len);
+            }
+        } catch (final Exception e) {
```

Comment:
Can you catch the specific exception (or reduce it to RuntimException). This comment applies to all locations that catch Exception.

Reply:
If there is an exception it means the thread is shutdown. Does this tickle down to the user in form of a status update offline? or does the binding simply stop working without visible indication and only by looking at the log one could find the error and deduce the binding has crashed.